pandas/tests/arithmetic/test_datetime64.py

```diff
@@ -2350,3 +2352,34 @@ def test_shift_months(years, months):
            for x in dti]
     expected = DatetimeIndex(raw)
     tm.assert_index_equal(actual, expected)
+
+
+def test_add_with_monkeypatched_datetime(monkeypatch):
+    # GH 25734
+
+    class MetaDatetime(type):
+        @classmethod
+        def __instancecheck__(self, obj):
+            return isinstance(obj, datetime)
+
+    class FakeDatetime(MetaDatetime("NewBase", (datetime,), {})):
+        pass
+
+    with monkeypatch.context() as m:
+        # monkeypatch datetime everywhere
+        for mod_name, module in list(sys.modules.items()):
```

Comment:
At least some level of monkeypatching is necessary for the original issue to show up.

However, this entire `for` block could be replaced with `m.setattr("pandas.tseries.offsets.datetime", FakeDatetime)`, but I wanted to future proof this test better and replicate what `freezegun` and other libraries that might monkeypatch `datetime` would do.